src/Qir/Runtime/lib/QIR/QubitManagerRestrictedReuse.hpp

```diff
@@ -0,0 +1,210 @@
+    // This class is small, contains no pointers and relies on default shallow copying/destruction.
+    struct RestrictedReuseArea final
+    {
+    public:
+        QubitListInSharedArray FreeQubitsReuseProhibited;
+        QubitListInSharedArray FreeQubitsReuseAllowed;
+
+        RestrictedReuseArea() = default;
+        RestrictedReuseArea(QubitListInSharedArray freeQubits);
+    };
+
+
+
+    // This is NOT a stack! We modify it only by push/pop, but we also iterate over elements.
+    // TODO: Better name?
+    class CRestrictedReuseAreaStack final : public std::vector<RestrictedReuseArea>
+    {
+    public:
+        // No complex scenarios for now. Don't need to support copying/moving.
+        CRestrictedReuseAreaStack() = default;
+        CRestrictedReuseAreaStack(const CRestrictedReuseAreaStack&) = delete;
+        CRestrictedReuseAreaStack& operator = (const CRestrictedReuseAreaStack&) = delete;
+        ~CRestrictedReuseAreaStack() = default;
+
+        void PushToBack(RestrictedReuseArea area);
+        RestrictedReuseArea PopFromBack();
+        RestrictedReuseArea& PeekBack();
+        // TODO: Remove and use size directly?
+        int Count();
+    };
+
+    class CQubitManagerRestrictedReuse : public IQubitManager
+    {
+        // Restricted reuse area control, allocation and release
+
+        public: void StartRestrictedReuseArea();
+        public: void NextRestrictedReuseSegment();
+        public: void EndRestrictedReuseArea();
+        private: long AllocateQubit(); // Computation complexity is O(number of nested restricted reuse areas).
+        private: void ReleaseQubit(long id);
+
+        // Configuration Properties
+
+        // TODO: Make sure these are read only
+        public: bool MayExtendCapacity;
+        public: bool EncourageReuse;
+
+        // Constructors/destructors, storage and reallocation
+
+        private: int* sharedQubitStatusArray; // Tracks allocation state of all qubits. Stores lists of free qubits.
+        private: int qubitCapacity = 0; // qubitCapacity is always equal to the array size.
+        private: CRestrictedReuseAreaStack freeQubitsInAreas; // Fresh Free Qubits are located in freeQubitsInAreas[0].FreeQubitsReuseAllowed
+        
+        public: CQubitManagerRestrictedReuse(
+            int initialQubitCapacity,
+            bool mayExtendCapacity = false,
+            bool encourageReuse = true);
+
+        // No complex scenarios for now. Don't need to support copying/moving.
+        public: CQubitManagerRestrictedReuse(const CQubitManagerRestrictedReuse&) = delete;
+        public: CQubitManagerRestrictedReuse& operator = (const CQubitManagerRestrictedReuse&) = delete;
+        public: ~CQubitManagerRestrictedReuse();
+        private: void EnsureCapacity(int requestedCapacity);
+
+        // Disable, disabled qubit count and checks
+
+        public: int DisabledQubitsCount = 0;
```

Comment:
FYI: In the C++ code I mostly saw, the member variables were typically grouped in a single `private` section in the end of the class. Such grouping helps to quickly estimate the total size of the class instance, and helps to check if all the member variables are initialized in the constructor.

Reply:
As I commented earlier this is the C# heritage. I rearranged the code now.